# ONT Shotgun Assembly\

## Assembly

Metagenomic assemblies can help you recover genes and pathways present in a community. They can also be fed into community analysis. On the other hand, you can miss some of the more rare members of the community.

Let's do an assembly of the Oxford Nanopore (ONT) reads from the red alder nodules. We will use the reads that have had red alder removed so we are only assembling the microbes. 

<!--Note that the nanoMDBG module assumes it is r10.4+ but based on the timing of the sequencing, the red alder is likely earlier.-->

We'll use metaMDBG, which is designed for long, accurate shotgun metagenomic reads including PacBio HiFi and Nanopore r10.4+. It's use of a minimizer de-Bruijn graph (MDBG) allows it to run quickly with low memory requirements. It can handle different coverages within a sample. For ONT data, it has nanoMDBG, by the same group, integrated.

https://github.com/GaetanBenoitDev/metaMDBG


Make sure you are in a screen.

```{bash,eval=FALSE}
screen -S assembly
```


Create a directory called nodule-assembly in your home directory. Then go into it.

```{bash,eval=FALSE}
mkdir ~/nodule-assembly
cd ~/nodule-assembly
```

Activate the environment.

```{bash,eval=FALSE}
conda activate metamdbg1.2
```

Note that there is something wrong with the metamdbg in the conda environment that we just activated but this will have the dependencies we need.

We will add the path to another metaMDBG installation to our $PATH variable so that linux knows where to find it.

<!--I tried to install in conda and there were some issues; leaving the conda env for now as I test the other install-->

```{bash,eval=FALSE}
export PATH=$PATH:/opt/metamdbg/metaMDBG/metaMDBG/build/bin
```

Assemble the metagenome. Note that you can assemble multiple samples together by feeding multiple fastq files in with the --in-ont flag. This will take ~20 minutes.

```{bash,eval=FALSE}
metaMDBG asm --out-dir 3469-3_assembly --in-ont ~/microbe_fastq/3469-3.microbe.fq.gz --threads 8
```


## Assembly Completeness and Contamination

We will feed in the whole assembly since it is mostly Frankia but later you will learn how to bin your assembly into the different organisms that it represents.

CheckM allows you to assess completeness and contamination of MAGs (Metagenome-assembled genomes). It is in the same environment (flye).

CheckM puts each of your sequences in a phylogenetic tree (if needed) and then looks for single-copy genes that should be in each sequence. The percentage of expected genes it finds is a measure of completeness. The number or multi-copy genes it finds are a measure of contamination (the presence of closely related or more distantly related organisms in your bin).

CheckM \
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4484387/pdf/1043.pdf \
https://ecogenomics.github.io/CheckM/


Here is a test output using 2 genomes from NCBI, a partial genome of one of them, and the 2 genomes catted together.

![CheckM Output](./Figures/checkm.png){width=100%}

Video explanation:  
https://onestopdataanalysis.com/checkm-completeness-contamination/


CheckM at the genus level. We will use taxonomy_wf (wf stands for workflow) which allows us to feed in the Frankia genus that we know is there.

checkm taxonomy_wf [parameters] <taxonomy_level> <taxon> <assembly_bin_dir> <output_dir>

The parameters we will use:

-x &nbsp;&nbsp;&nbsp;&nbsp; extension (extension of the assembly bins, ie fasta;   checkm will run on all the files in the assembly_bin_dir that have that extension)  
-t &nbsp;&nbsp;&nbsp;&nbsp; number of threads  
-f &nbsp;&nbsp;&nbsp;&nbsp; output file (default: STDOUT)

For additional parameters:
```{bash,eval=FALSE}
checkm taxonomy_wf --help
```

```{bash,eval=FALSE}
cd 3469-3_assembly

checkm taxonomy_wf -x fasta -t 8 -f ./3469-3_tax_results.txt genus Frankia ~/nodule-assembly-3469-3 ./checkm
```

Note: You can also use the lineage workflow (lineage_wf) if you don't know your taxa. It will place sequences on a tree and calculate their lineages.

Plot the genome completeness, contamination and strain heterogeneity stats.

checkm marker_plot [parameters] <results_dir> <assembly_bin_dir> <output_dir>

The parameters we will use:  

-x &nbsp;&nbsp;&nbsp;&nbsp; extension (extension of the assembly bins, ie fasta; checkm will run on all the files in the assembly_bin_dir that have that extension)  
--image_type &nbsp;&nbsp;&nbsp;&nbsp; format of image (eps,pdf,png,ps,or svg; default: png)  
--dpi <dpi> &nbsp;&nbsp;&nbsp;&nbsp; dots per inch resolution for output file (default: 600)  
--font_size <size> &nbsp;&nbsp;&nbsp;&nbsp; font size (default: 8)  
--height <height> &nbsp;&nbsp;&nbsp;&nbsp; height in inches (default: 6.5)

```{bash,eval=FALSE}
checkm marker_plot -x fasta --image_type png --dpi 600 --font_size 14 --height 5 \
~/nodule-assembly-3469-3/checkm ~/nodule-assembly-3469-3 ~/nodule-assembly-3469-3/checkm
```


Let's look at the assembly metrics generated by checkM.

```{bash,eval=FALSE}
less ~/nodule-assembly-3469-3/checkm/storage/bin_stats.analyze.tsv
```

Deactivate the environment.

```{bash,eval=FALSE}
conda deactivate
```

Download the marker_plot to your computer using filezilla and take a look at it.


Note: You can also classify the assembly contigs using KrakenUniq or Centrifuge. Some of the rare organisms will have some reads but not enough to assemble, so you might lose some of the rarer taxa.

### Strengths and weaknesses of assembly-based and read-based metagenomics analysis

![Quince, C., Walker, A. W., Simpson, J. T., Loman, N. J., & Segata, N. (2017). Shotgun metagenomics, from sampling to analysis. Nature biotechnology, 35(9), 833-844.](./Figures/AssemblyvsRead.png){width=100%}


