<!â€”The pdf from the summer workshop has instructions on classifying reads and assembly contigs with Centrifuge. We won't do it here since they will learn Centrifuge with the Illumina data and we have already done the reads with krakenuniq.-->

# Nanopore WGS Assembly (metaflye)\

## Assembly

Metagenomic assemblies can help you recover genes and pathways present in a community. They can also be fed into community analysis. On the other hand, you can miss some of the more rare members of the community.

Let's do an assembly of the Oxford Nanopore (ONT) reads from the red alder nodules. We will use the reads that have had red alder removed so we are only assembling the microbes. We'll use the metagenomics module of the Flye assembler.

Create a directory called nodule-assembly in your home directory. Then go into it.

```
mkdir ~/nodule-assembly
cd ~/nodule-assembly
```

Create a soft link to the Red Alder genome.

```
ln -s /home/data/metagenomics-2310/red-alder-reads/red-alder-genome.fasta .
```

Activate the environment.

```
conda activate flye
```

Assemble the metagenome with metaFlye. Don't forget the dot. That tells the program to put the output in the current directory.

```
flye --nano-raw ~/microbe_fastq/3469-3.microbe.fq -i 1 --meta -t 8 -g 45m -o .
```

Assess the Quality of the Root Nodule Metagenome Assemblies with CheckM. It is in the same environment (flye).

CheckM at the genus level

```
checkm taxonomy_wf genus Frankia ~/nodule-assembly ./checkm -x fasta -t 8 -f ./3469-3_tax_results.txt
```

Plot the genome completeness, contamination and strain heterogeneity stats.

```
checkm marker_plot -x fasta --image_type png --dpi 600 --font_size 14 --height 5 \
~/nodule-assembly/checkm ~/nodule-assembly ~/nodule-assembly/checkm
```

Let's look at the assembly metrics generated by checkM.

```
less ~/nodule-assembly/checkm/storage/bin_stats.analyze.tsv
```

Download the marker_plot to your computer using scp, moba-xterm, or filezilla.


Deactivate the environment.

```
conda deactivate flye
```

Note: You can classify the nanopore microbial reads or the assembly contigs using Centrifuge (community analysis). Note that we have already classified the reads with krakenuniq and the assemblies with checkM. You will learn how to use centrifuge when we work on Illumina whole genome (shotgun) data so we won't do it here.
