# Graph-specific Analyses

![Analysis](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/GraphSpecific.png){width=50%} 

## Frequented Region (FR) Algorithm

Identify regions in a graph that are co-visited by a set of supporting
paths from individual sequences in the pangenome.

![FR](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/FRAlg.png){width=80%} 


1. Similar evolutionary histories
2. Similar selection pressures

3. Heuristic but scalable
4. Node merging with noise tolerance
    + Error
    + Divergence
5. Nucleotide level

### A few of the parameters

1. penetrance (ùõº)
minimum fraction of nodes
2. the maximum insertion parameter (Œ∫)
maximum number of inserted nodes

## Yeast Insertion for Alcohol Tolerance

### 17kb insertion on chromosome XIV

![Alcohol Tolerance Insertion](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/etohTolerance.png){width=100%} 

### 17kb insertion across all 55 strains

+ 9 of 20 additional alcohol related strains share EC1118‚Äôs inserted FRs. 
+ Strains sharing four or more FRs are highlighted.
+ 5 genes inserted


![Insertion](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Insertion.png){width=100%} 

## SARS-CoV-2

+ Identify conserved genomic regions
    + Vaccine and treatment targets

+ Identify genomic shifts over time and geographic space
    + Mutations affecting spread and virulence

+ Identify genomic regions that differentiate strains
    + Implications in modelling and treatment
 
![Virus Cartoon](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Virus.png){width=20%}

## Genomic shifts across time and space

733 SARS-CoV-2 genomes from the beginning of the pandemic through April

Highlight

+ Early strains:
    + (Wuhan, 12/30/2019)
    + (Italy, 4/27/2020)
    
+ **Spike protein**
    + Engages the human ACE2 cell receptor to gain entry into the host cell. Target for vaccines, template for antibody production, and candidate for generating entry inhibitors.
    
### Wuhan & Italy compared to 733 strain support

Spike protein conserved, especially end Genomic shift

![Shift](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Shift.png){width=100%}

![Path Support](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/PathSupport.png){width=100%}

### D614G mutation

A mutation seen in Europe in February that spread across the globe to become the dominant strain. Preliminary analyses suggest it might increase contagion and/or virulence and that strains people with immunity to the original strain might not be immune to the mutated strain. The figure shows the region has undergone genomic shift and isn‚Äôt conserved in all the genomes.

![D614G](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/D614G.png){width=100%}

### Conserved haplotype blocks and selection coefficients

##### Soft and incomplete selective sweeps{-}

![Haplotype Blocks](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/haplotype.png){width=80%}



![ORF1a](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Orf1a.png){width=100%}

## Medicago truncatula

![M. truncatula](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Medicago.png){width=50%}

+ Model legume
+ 450 Mb genome
+ Selfing
+ Whole genome duplications
+ High level of rearrangements and gene family expansions

### Medicago Pan-13 genome size curve

![M. truncatula genome](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Medicago2.png){width=100%}

+ Zhou, P., Silverstein, K.A., Ramaraj, T., Guhlin, J., Denny, R., Liu,
J., Farmer, A.D., Steele, K.P., Stupar, R.M. & Miller, J.R. Exploring
structural variation and gene family architecture with De Novo
assemblies of 15 Medicago genomes. BMC genomics 18, 261 (2017).

### Medicago Pan-13 *proteome* size curve

![M. truncatula proteome](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/MedicagoProteome.png){width=100%}

### Medicago HapMap Summary

1. As much as 22% of the genome is involved in large structural changes;
2. Medicago 13 pangenome expands reference space by 16% (63 Mbp);
3. Pangenome analysis revealed that 42% (180 Mbp) of genomic sequences are missing in one or more assemblies, while pan-proteome analysis identified 67% (50,700) of all ortholog groups as dispensable;
4. Environmentally Sensitive Gene Families found to be enriched in the accession-specific gene pool.

### Medicago: 23 accessions for de novo assembly 

1. 15 genomes from 3 phylogenetic hubs (incl. A17)
2. 6 genomes from RIL parents
3. 3 ‚Äúrandom‚Äù genomes

![Accessions](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Accessions.png){width=60%}

### Medicago: LEED..PEED (LP) family

1. Short secreted peptides
2. Transcribed in nitrogen-fixing nodules but absent from other plant tissues
3. Not found in other sequenced legumes
4. M. truncatula A17 (reference)
    + 12 LP genes on chromosome 7
    + 1 LP gene on chromosome 4
    
+ leucine:glutamic acid:glutamic acid:aspartic acid
+ proline:glutamic acid:glutamic acid:aspartic acid

Genomic Characterization of the LEED..PEEDs, a Gene Family Unique to the Medicago Lineage
Diana I. Trujillo, Kevin A. T. Silverstein and Nevin D. Young 
G3: GENES, GENOMES, GENETICS October 1, 2014 vol. 4 no. 10 2003-2012; https://doi.org/10.1534/g3.114.011874

#### LEED..PEED{-}

![LEED..PEED](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/LeedPeed.png){width=100%}

### *Medicago truncatula*

1. Horizontal gene transfer among symbionts and between symbionts and plants

2. Nodule specific gene families

3. Expansion of plant gene families

## Using Machine Learning to Connect Genotype to Phenotype

![Predicting Phenotype](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/GWAS.png){width=100%}

### Yeast

1. FR matrix as features
2. Random Forest
3. Predict Phenotypes

### Yeast Root Mean Square Error (RMSE)

![RMSE](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/RMSE.png){width=100%}

![RMSE](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/RMSE2.png){width=100%}

![Predicted Phenotype](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Phenotype.png){width=100%}

## VG and Machine Learning

### vg vectorize (xg graph or gcsa)

![Predicted Phenotype](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Vectorize1.png){width=100%}

![Predicted Phenotype](~/Desktop/Bookdown/Pangenomics-Bookdown/Figures/Vectorize2.png){width=100%}




