<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>QIIME2</title>
  <meta name="description" content="QIIME2" />
  <meta name="generator" content="bookdown 0.33 and GitBook 2.6.7" />

  <meta property="og:title" content="QIIME2" />
  <meta property="og:type" content="book" />
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="QIIME2" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  


<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Pangenomics Workshop NCGR</a></li>

<li class="divider"></li>
<li class="chapter" data-level="0.1" data-path=""><a href="#logging-on-to-the-server"><i class="fa fa-check"></i><b>0.1</b> Logging on to the server</a></li>
<li class="chapter" data-level="0.2" data-path=""><a href="#setting-up-a-working-directory-for-qiime2-analysis"><i class="fa fa-check"></i><b>0.2</b> Setting up a working directory for QIIME2 analysis</a></li>
<li class="chapter" data-level="0.3" data-path=""><a href="#dataset-for-analysis"><i class="fa fa-check"></i><b>0.3</b> Dataset for analysis</a></li>
<li class="chapter" data-level="0.4" data-path=""><a href="#data-download"><i class="fa fa-check"></i><b>0.4</b> Data download</a></li>
<li class="chapter" data-level="0.5" data-path=""><a href="#quality-control"><i class="fa fa-check"></i><b>0.5</b> Quality Control</a></li>
<li class="chapter" data-level="0.6" data-path=""><a href="#what-is-qiime2"><i class="fa fa-check"></i><b>0.6</b> What is QIIME2?</a></li>
<li class="chapter" data-level="0.7" data-path=""><a href="#metadata-setup"><i class="fa fa-check"></i><b>0.7</b> Metadata setup</a></li>
<li class="chapter" data-level="0.8" data-path=""><a href="#data-import"><i class="fa fa-check"></i><b>0.8</b> Data import</a></li>
<li class="chapter" data-level="0.9" data-path=""><a href="#dereplicate"><i class="fa fa-check"></i><b>0.9</b> Dereplicate</a></li>
<li class="chapter" data-level="0.10" data-path=""><a href="#generating-a-feature-table"><i class="fa fa-check"></i><b>0.10</b> Generating a feature table</a></li>
<li class="chapter" data-level="0.11" data-path=""><a href="#visualization"><i class="fa fa-check"></i><b>0.11</b> Visualization</a></li>
<li class="chapter" data-level="0.12" data-path=""><a href="#setting-up-a-reference"><i class="fa fa-check"></i><b>0.12</b> Setting up a reference</a></li>
<li class="chapter" data-level="0.13" data-path=""><a href="#clustering"><i class="fa fa-check"></i><b>0.13</b> Clustering</a></li>
<li class="chapter" data-level="0.14" data-path=""><a href="#pcoa-plot"><i class="fa fa-check"></i><b>0.14</b> PCoA plot</a></li>
<li class="chapter" data-level="0.15" data-path=""><a href="#assigning-taxonomy"><i class="fa fa-check"></i><b>0.15</b> Assigning taxonomy</a></li>
<li class="chapter" data-level="0.16" data-path=""><a href="#visualizing"><i class="fa fa-check"></i><b>0.16</b> Visualizing</a></li>
<li class="chapter" data-level="0.17" data-path=""><a href="#demultiplexing"><i class="fa fa-check"></i><b>0.17</b> Demultiplexing</a></li>
<li class="chapter" data-level="0.18" data-path=""><a href="#denoising"><i class="fa fa-check"></i><b>0.18</b> Denoising</a></li>
<li class="chapter" data-level="0.19" data-path=""><a href="#chimeric-feature-filtering"><i class="fa fa-check"></i><b>0.19</b> Chimeric feature filtering</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">QIIME2</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">QIIME2</h1>
<h2 class="subtitle"><em>Hands On Command Line Tutorial</em></h2>
</div>
<div id="logging-on-to-the-server" class="section level2 hasAnchor" number="0.1">
<h2><span class="header-section-number">0.1</span> Logging on to the server<a href="#logging-on-to-the-server" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Make it a practice to start a screen before you begin analysis.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ssh</span> <span class="at">-p</span> 44111 <span class="op">&lt;</span>USERNAME<span class="op">&gt;</span>@gateway.training.ncgr.org</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">screen</span> <span class="at">-S</span> QIIME2</span></code></pre></div>
</div>
<div id="setting-up-a-working-directory-for-qiime2-analysis" class="section level2 hasAnchor" number="0.2">
<h2><span class="header-section-number">0.2</span> Setting up a working directory for QIIME2 analysis<a href="#setting-up-a-working-directory-for-qiime2-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/<span class="va">$USER</span>/</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> QIIME2_Analysis</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> QIIME2_Analysis</span></code></pre></div>
</div>
<div id="dataset-for-analysis" class="section level2 hasAnchor" number="0.3">
<h2><span class="header-section-number">0.3</span> Dataset for analysis<a href="#dataset-for-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For this tutorial, we will use data from the following published manuscript:
<a href="https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-29" class="uri">https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-29</a></p>
<p>Respiratory tract clinical sample selection for microbiota analysis in patients with pulmonary tuberculosis</p>
<p>Pulmonary Tuberculosis (The Human Microbiome in the Fight Against Tuberculosis, 2017)</p>
<p><img src="QIIMEpics/lyx.qiime.png" style="width:100.0%" alt="tuberculosis" />
Why study microbiome in relation to TB?
1. Microbiome composition is a potential risk factors for TB infection and disease susceptibility, disease progression,
and treatment outcomes
2. TB infection and disease can affect microbiome, subsequently impacting health via alteration of immune responses</p>
<p>This pilot study was carried out using sputum, oropharynx, and nasal respiratory tract samples collected from patients
with pulmonary tuberculosis and healthy control individuals, in order to compare sample types and their usefulness in
assessing changes in bacterial and fungal communities.</p>
</div>
<div id="data-download" class="section level2 hasAnchor" number="0.4">
<h2><span class="header-section-number">0.4</span> Data download<a href="#data-download" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Data was downloaded from the National Center for Biotechnology Information (NCBI) based on the SRA accession
numbers reported in the manuscript - PRJNA242354.
A program called fastq-dump version 2.10.0 was used to download the data directly from NCBI SRA database. Tbe
following exercise will help you use fastq-dump for your own projects too.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate fastqdump</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># MAKE SURE YOU ARE IN THE FOLLOWING FOLDER</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">/home/</span><span class="va">$USER</span><span class="ex">/QIIME2_Analysis</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># CREATE A FOLDER FOR DOWNLOADING FASTQ FILES TO ANALYZE</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> input_data</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> input_data</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="ex">fastq-dump</span> <span class="at">--gzip</span> <span class="at">--outdir</span> ./ <span class="dt">\</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>SRR1204628 SRR1204630 SRR1204627 SRR1204629 SRR1204626 SRR1204623 <span class="dt">\</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>SRR1204625 SRR1204624 SRR1204622 SRR1204631 SRR1204638 SRR1204642 <span class="dt">\</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>SRR1204633 SRR1204632 SRR1204634 SRR1204639 SRR1204646 SRR1204647 <span class="dt">\</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>SRR1204641 SRR1204640 SRR1204636 SRR1204637 SRR1369696 SRR1369694 <span class="dt">\</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>SRR1204635 SRR1204643 SRR1204644 SRR1204645 SRR1369709 SRR1369707 <span class="dt">\</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>SRR1369690 SRR1369691 SRR1369692 SRR1369693 SRR1369710 SRR1369713 <span class="dt">\</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>SRR1204650 SRR1204651 SRR1369688 SRR1369689 SRR1369695 SRR1204649 <span class="dt">\</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>SRR1204648 SRR1369697 SRR1369704 SRR1369703 SRR1369706 SRR1369705 <span class="dt">\</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>SRR1369708 SRR1369698 SRR1369699 SRR1369701 SRR1369700 SRR1369702 <span class="dt">\</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>SRR1369714 SRR1369717 SRR1369716 SRR1369715 SRR1369711 SRR1369712</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="co"># --outdir -&gt; Path to Output directory</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="co"># --gzip -&gt; Compress output files using gzip</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a><span class="co"># After running this step, you will have 60 files in your input_data folder</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a><span class="co"># To check, cd into folder that has fastq files, type this command</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Optional exercise:</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Count the number of sequences in the two fastq files pertaining to one sample and sum them.</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre></div>
</div>
<div id="quality-control" class="section level2 hasAnchor" number="0.5">
<h2><span class="header-section-number">0.5</span> Quality Control<a href="#quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter poor quality reads</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate fastp</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> <span class="pp">*</span>.fastq<span class="kw">;</span> <span class="cf">do</span> <span class="ex">fastp</span> <span class="at">-i</span> <span class="va">${file}</span> <span class="at">-o</span> <span class="va">${file}</span>.filtered<span class="kw">;</span> <span class="cf">done</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span></code></pre></div>
<p><img src="QIIMEpics/lyx.qiime.table.png" style="width:100.0%" alt="tuberculosis" />
Data was collected from 30 Individuals.</p>
<p>• 6 TB Patients - Nasal</p>
<p>• 6 TB Patients - oropharynx</p>
<p>• 6 TB Patients - Sputum</p>
<p>• 6 Healthy Controls - Nasal</p>
<p>• 6 Healthy Controls - oropharynx</p>
<p>DATA TABLE:</p>
<pre><code>SAMPLE FASTA FILE                   TOTAL NUMBER OF READS 
Control_1_Nasal_L001_R1_001.fasta   41056
Control_1_Pharynx_L001_R1_001.fasta 119972
Control_2_Nasal_L001_R1_001.fasta   32142
Control_2_Pharynx_L001_R1_001.fasta 42973
Control_3_Nasal_L001_R1_001.fasta   32953
Control_3_Pharynx_L001_R1_001.fasta 39148
Control_4_Nasal_L001_R1_001.fasta   83174
Control_4_Pharynx_L001_R1_001.fasta 17475
Control_5_Nasal_L001_R1_001.fasta   29146
Control_5_Pharynx_L001_R1_001.fasta 32728
Control_6_Nasal_L001_R1_001.fasta   32738
Control_6_Pharynx_L001_R1_001.fasta 15320
Patient_1_Nasal_L001_R1_001.fasta   69995
Patient_1_Pharynx_L001_R1_001.fasta 59646
Patient_1_Sputum_L001_R1_001.fasta  66725
Patient_2_Nasal_L001_R1_001.fasta   61070
Patient_2_Pharynx_L001_R1_001.fasta 27106
Patient_2_Sputum_L001_R1_001.fasta  71434
Patient_3_Nasal_L001_R1_001.fasta   91920
Patient_3_Pharynx_L001_R1_001.fasta 36224
Patient_3_Sputum_L001_R1_001.fasta  27533
Patient_4_Nasal_L001_R1_001.fasta   28996
Patient_4_Pharynx_L001_R1_001.fasta 30713
Patient_4_Sputum_L001_R1_001.fasta  59433
Patient_5_Nasal_L001_R1_001.fasta   70843
Patient_5_Pharynx_L001_R1_001.fasta 39674
Patient_5_Sputum_L001_R1_001.fasta  29576
Patient_6_Nasal_L001_R1_001.fasta   106178
Patient_6_Pharynx_L001_R1_001.fasta 58956
Patient_6_Sputum_L001_R1_001.fasta  35703</code></pre>
</div>
<div id="what-is-qiime2" class="section level2 hasAnchor" number="0.6">
<h2><span class="header-section-number">0.6</span> What is QIIME2?<a href="#what-is-qiime2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://docs.qiime2.org/2020.11/about/" class="uri">https://docs.qiime2.org/2020.11/about/</a></p>
<p>QIIME 2 is a powerful microbiome analysis package with a focus on data and analysis transparency. QIIME2 enables researchers to start an analysis with raw DNA sequence data and finish with publication-quality figures and statistical results.</p>
<p>Key features:</p>
<p>• Integrated and automatic tracking of data provenance</p>
<p>• Plugin system for extending microbiome analysis functionality</p>
<p>• Support for multiple types of user interfaces (e.g. API, command line, graphical)</p>
<p>Plugin availability - <a href="https://docs.qiime2.org/2020.11/plugins/available/" class="uri">https://docs.qiime2.org/2020.11/plugins/available/</a></p>
</div>
<div id="metadata-setup" class="section level2 hasAnchor" number="0.7">
<h2><span class="header-section-number">0.7</span> Metadata setup<a href="#metadata-setup" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Metadata file that we will be using for our data set: (mappingFile.txt on logrus)</p>
<p>Here is how you will make a metadata file on your own: <a href="https://docs.qiime2.org/2017.6/tutorials/metadata/" class="uri">https://docs.qiime2.org/2017.6/tutorials/metadata/</a></p>
<pre><code>#SampleID   Treatment   TreatmentSite   TreatmentNumber     Sex     BodySite    Description 
C_1_N   C   CN  1   F   Nasal       Control 1 Nasal 
C_1_P   C   CP  2   F   Oropharynx  Control 1 Pharynx 
C_2_N   C   CN  1   M   Nasal       Control 2 Nasal 
C_2_P   C   CP  2   M   Oropharynx  Control 2 Pharynx 
C_3_N   C   CN  1   M   Nasal       Control 3 Nasal 
C_3_P   C   CP  2   M   Oropharynx  Control 3 Pharynx 
C_4_N   C   CN  1   M   Nasal       Control 4 Nasal 
C_4_P   C   CP  2   M   Oropharynx  Control 4 Pharynx 
C_5_N   C   CN  1   M   Nasal       Control 5 Nasal 
C_5_P   C   CP  2   M   Oropharynx  Control 5 Pharynx 
C_6_N   C   CN  1   M   Nasal       Control 6 Nasal 
C_6_P   C   CP  2   M   Oropharynx  Control 6 Pharynx 
P_1_N   P   PN  3   F   Nasal       Patient 1 Nasal 
P_1_P   P   PP  4   F   Oropharynx  Patient 1 Pharynx 
P_1_S   P   PS  5   F   Sputum      Patient 1 Sputum 
P_2_N   P   PN  3   M   Nasal       Patient 2 Nasal 
P_2_P   P   PP  4   M   Oropharynx  Patient 2 Pharynx 
P_2_S   P   PS  5   M   Sputum      Patient 2 Sputum 
P_3_N   P   PN  3   M   Nasal       Patient 3 Nasal 
P_3_P   P   PP  4   M   Oropharynx  Patient 3 Pharynx 
P_3_S   P   PS  5   M   Sputum      Patient 3 Sputum 
P_4_N   P   PN  3   M   Nasal       Patient 4 Nasal 
P_4_P   P   PP  4   M   Oropharynx  Patient 4 Pharynx 
P_4_S   P   PS  5   M   Sputum      Patient 4 Sputum 
P_5_N   P   PN  3   M   Nasal       Patient 5 Nasal 
P_5_P   P   PP  4   M   Oropharynx  Patient 5 Pharynx 
P_5_S   P   PS  5   M   Sputum      Patient 5 Sputum 
P_6_N   P   PN  3   M   Nasal       Patient 6 Nasal 
P_6_P   P   PP  4   M   Oropharynx  Patient 6 Pharynx 
P_6_S   P   PS  5   M   Sputum      Patient 6 Sputum </code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Move out of the input_data directory</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> ../</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy metadata file from my workspace to your workspace,</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/<span class="va">$USER</span>/QIIME2_Analysis</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> /home/elavelle/QIIME2_Analysis/mappingFile.txt ./</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">less</span> mappingFile.txt </span></code></pre></div>
</div>
<div id="data-import" class="section level2 hasAnchor" number="0.8">
<h2><span class="header-section-number">0.8</span> Data import<a href="#data-import" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We will be beginning our analysis with demultiplexed, quality controlled sequences.</p>
<p>For the purpose of this workshop we will not use the entire dataset per sample. We have created 1000 sequences per sample as subset. This will be your input dataset.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /home/<span class="va">$USER</span>/QIIME2_Analysis</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the subset file of sequences:</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> /home/metag/data/seqs.1000.fna ./</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> deactivate</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate qiime2</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># This command will vary depending input data type and format</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Here, we are importing .fasta reads (no quality scores)</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> tools import <span class="dt">\</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>--input-path seqs.1000.fna <span class="dt">\</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>--output-path seqs.1000.qza <span class="dt">\</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>--type <span class="st">&#39;SampleData[Sequences]&#39;</span></span></code></pre></div>
</div>
<div id="dereplicate" class="section level2 hasAnchor" number="0.9">
<h2><span class="header-section-number">0.9</span> Dereplicate<a href="#dereplicate" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>From <a href="https://docs.qiime2.org/2021.2/tutorials/overview/#derep-denoise" class="uri">https://docs.qiime2.org/2021.2/tutorials/overview/#derep-denoise</a></p>
<p>• We dereplicate our sequences to reduce repetition and file size/memory requirements in downstream steps (Don’t worry, we keep count of each replicate).</p>
<p>• We denoise to correct for stochastic sequencing effects and PCR error.</p>
<p>• We cluster sequences to collapse similar sequences (e.g., those that are ≥ 97% similar to each other) into single replicate sequences. This process, also known as OTU picking, was once a common procedure, used to simultaneously dereplicate but also perform a quick-and-dirty denoising procedure. Use denoising methods instead if you can.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> vsearch dereplicate-sequences <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>--i-sequences seqs.1000.qza <span class="dt">\</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>--o-dereplicated-table table.qza <span class="dt">\</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>--o-dereplicated-sequences rep-seqs.qza </span></code></pre></div>
</div>
<div id="generating-a-feature-table" class="section level2 hasAnchor" number="0.10">
<h2><span class="header-section-number">0.10</span> Generating a feature table<a href="#generating-a-feature-table" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The final products of all denoising and clustering methods/workflows are a FeatureTable[Frequency] (feature table) artifact and a FeatureData[Sequence] (representative sequences) artifact. These are two of the most important artifacts in an amplicon sequencing workflow, and are used for many downstream analyses, as discussed below. Indeed, feature tables are crucial to any QIIME 2 analysis, as the central record of all observations per sample.</p>
<p><img src="derep-denoise.png" style="width:100.0%" alt="workflow" />
The feature-table summarize command will give you information on how many sequences are associated with each sample and with each feature, histograms of those distributions, and some related summary statistics. The feature-table and tabulate-seqs commands will provide a mapping of feature IDs to sequences, and provide links to easily BLAST each sequence against the NCBI nt database.</p>
</div>
<div id="visualization" class="section level2 hasAnchor" number="0.11">
<h2><span class="header-section-number">0.11</span> Visualization<a href="#visualization" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-table summarize <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>--o-visualization table.qzv</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-table tabulate-seqs <span class="dt">\</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>--i-data rep-seqs.qza <span class="dt">\</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>--o-visualization rep-seqs.qzv</span></code></pre></div>
<p>Look at the format of a feature table from a mock dataset as well as other tutorials:
<a href="https://view.qiime2.org/visualization/?type=html&amp;src=https%3A%2F%2Fdocs.qiime2.org%2F2020.11%2Fdata%2Ftutorials%2Fmoving-pictures%2Ftable.qzv" class="uri">https://view.qiime2.org/visualization/?type=html&amp;src=https%3A%2F%2Fdocs.qiime2.org%2F2020.11%2Fdata%2Ftutorials%2Fmoving-pictures%2Ftable.qzv</a></p>
</div>
<div id="setting-up-a-reference" class="section level2 hasAnchor" number="0.12">
<h2><span class="header-section-number">0.12</span> Setting up a reference<a href="#setting-up-a-reference" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Greengenes (16S rRNA) and Silva (16S/18S rRNA) databases have to be imported as QIIME2 artifacts before using them.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> tools import <span class="dt">\</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>--input-path /home/metag/QIIME2_greengenes/Greengenes_16S_rRNA/<span class="dt">\</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>gg_13_8_otus/rep_set/99_otus.fasta <span class="dt">\</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>--output-path 99_otus.qza <span class="dt">\</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>--type FeatureData[Sequence]</span></code></pre></div>
</div>
<div id="clustering" class="section level2 hasAnchor" number="0.13">
<h2><span class="header-section-number">0.13</span> Clustering<a href="#clustering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><img src="QIIMEpics/qiimeimage.png" style="width:100.0%" alt="workflow" />
There are three types of clustering,</p>
<p>I. De novo Operational Taxonomic Unit (OTU) clustering</p>
<p>De novo OTU picking is used when we dont have a reference sequence collection to cluster against, for example, working with an infrequently used marker gene.</p>
<p>De novo OTU picking cannot be used in the following two scenarios,</p>
<ol style="list-style-type: decimal">
<li><p>Comparing non-overlapping amplicons, such as V2 and V4 regions of 16S rRNA</p></li>
<li><p>Working with very large data sets, e.g., HiSeq</p></li>
</ol>
<p>Advantage: No reference is required.</p>
<p>Disadvantage: Run time can be very expensive depending on the volume of data (parallel processing not possible)</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXAMPLE- DO NOT RUN THE ABOVE COMMAND TO COMPLETION.</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> vsearch cluster-features-de-novo <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>--i-sequences rep-seqs.qza <span class="dt">\</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>--p-perc-identity 0.99 <span class="dt">\</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>--o-clustered-table table-dn-99.qza <span class="dt">\</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>--o-clustered-sequences rep-seqs-dn-99.qza</span></code></pre></div>
<ol start="2" style="list-style-type: upper-roman">
<li>Closed-reference OTU clustering</li>
</ol>
<p>Reads are clustered against a reference sequence collection. Any reads which do not hit a sequence in the reference sequence collection are excluded from downstream analyses.</p>
<p>Closed-reference OTU picking is used:</p>
<ol style="list-style-type: decimal">
<li><p>When comparing non-overlapping amplicons, such as V2 and V4 regions of the 16S rRNA.</p></li>
<li><p>When reference sequences span the sequenced regions.</p></li>
</ol>
<p>Closed-reference OTU picking is not used when reference sequence collection is not available</p>
<p>Advantage: Fast, useful for extremely large data sets. Better trees and taxonomy. OTUs are defined in the reference and have a reliable tree/taxonomy.</p>
<p>Disadvantage: Inability to detect novel diversity with respect to your reference sequence collection</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># EXAMPLE- DO NOT RUN THE ABOVE COMMAND TO COMPLETION</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> vsearch cluster-features-closed-reference <span class="dt">\</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>--i-sequences rep-seqs.qza <span class="dt">\</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>--i-reference-sequences 99_otus.qza <span class="dt">\</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>--p-perc-identity 0.99 <span class="dt">\</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>--o-clustered-table table-cr-99.qza <span class="dt">\</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>--o-clustered-sequences rep-seqs-cr-99.qza <span class="dt">\</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>--o-unmatched-sequences unmatched-cr-99.qza</span></code></pre></div>
<ol start="3" style="list-style-type: upper-roman">
<li>Open-reference OTU clustering</li>
</ol>
<p>In open-reference OTU clustering, sequence reads are clustered against a reference sequence collection and any reads which do not hit the reference sequence collection are subsequently clustered de novo.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># EXAMPLE- DO NOT RUN THE ABOVE COMMAND TO COMPLETION</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> vsearch cluster-features-open-reference <span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>--i-sequences rep-seqs.qza <span class="dt">\</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>--i-reference-sequences 99_otus.qza <span class="dt">\</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>--p-perc-identity 0.99 <span class="dt">\</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>--o-clustered-table table-or-99.qza <span class="dt">\</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>--o-clustered-sequences rep-seqs-or-99.qza <span class="dt">\</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>--o-new-reference-sequences new-ref-seqs-or-99.qza</span></code></pre></div>
<p>STEP 7: Building a phylogenetic tree</p>
<p>Evolution is the core concept of biology. There is only so much you can learn from microbes while ignoring evolution. Evolution-aware analyses of a dataset needs a phylogenetic tree of its sequences. These trees can be built de-novo or they can be reference based. Phylogeny-based analyses in QIIME2 generally need a rooted tree.</p>
<div class="figure">
<img src="QIIMEpics/forlyx.trees.qiime.png" style="width:100.0%" alt="" />
<p class="caption">workflow</p>
</div>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Carry out a multiple sequence alignment using Mafft</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> alignment mafft <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>--i-sequences rep-seqs-or-99.qza <span class="dt">\</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>--o-alignment aligned-rep-seqs.qza</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 4 MINUTES TO RUN TO COMPLETION</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Mask (or filter) the alignment to remove positions that are highly variable. </span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># These positions are generally considered to add noise to a</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># resulting phylogenetic tree.</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> alignment mask <span class="dt">\</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>--i-alignment aligned-rep-seqs.qza <span class="dt">\</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>--o-masked-alignment masked-aligned-rep-seqs.qza</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 4 MINUTES TO RUN TO COMPLETION</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the tree using the Fasttree program</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> phylogeny fasttree <span class="dt">\</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>--i-alignment masked-aligned-rep-seqs.qza <span class="dt">\</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>--o-tree unrooted-tree.qza</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 9 MINUTES TO RUN TO COMPLETION</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a><span class="co"># root the tree using the longest root</span></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> phylogeny midpoint-root <span class="dt">\</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>--i-tree unrooted-tree.qza <span class="dt">\</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>--o-rooted-tree rooted-tree.qza</span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 6 SECONDS TO RUN TO COMPLETION</span></span></code></pre></div>
<p>STEP 8: Alpha rarefaction</p>
<p>What is rarefaction?</p>
<p>Species richness is the number of species present in a system. However, the deeper you sequence, the more species you will find. That can be problematic if you gathered 500 reads from one sample and only 100 reads from another sample- you would expect to find more species if you sequenced 5x as many reads! To account for this, we perform an in-silico (i.e., on your computer) experiment called rarefaction. A rarefaction is a random collection of sequences from a sample, with a specified depth (number of sequences). For example, a rarefaction with a depth of 75 reads per sample is a simulation of what your sequencing results would look like if you sequenced exactly 75 reads from each sample. To look at alpha diversity systematically, we can perform many rarefactions: at multiple depths and repeat many times at each depth. In QIIME, this task is performed on your OTU table.</p>
<p>• Concerns:</p>
<p>– Too low: ignore a lot of samples’ information</p>
<p>– Too high: ignore a lot of samples</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># EXAMPLE- DO NOT RUN THE ABOVE COMMAND TO COMPLETION</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cp</span> /home/elavelle/mappingFile.txt ./</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity alpha-rarefaction <span class="dt">\</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>--i-table table-or-99.qza <span class="dt">\</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>--i-phylogeny rooted-tree.qza <span class="dt">\</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>--p-max-depth 1000 <span class="dt">\</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>--o-visualization alpha-rarefaction.qzv</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Why specify a maximum depth of 1000?</span></span></code></pre></div>
<p>Look at the following example and decide what sampling-depth you would use:</p>
<p><a href="https://view.qiime2.org/visualization/?type=html&amp;src=https%3A%2F%2Fdocs.qiime2.org%2F2020.11%2Fdata%2Ftutorials%2Fmoving-pictures%2Falpha-rarefaction.qzv" class="uri">https://view.qiime2.org/visualization/?type=html&amp;src=https%3A%2F%2Fdocs.qiime2.org%2F2020.11%2Fdata%2Ftutorials%2Fmoving-pictures%2Falpha-rarefaction.qzv</a></p>
<p>STEP 9: Calculating and exploring diversity metrics</p>
<p>Species diversity is a valuable tool for describing the ecological complexity of a single sample (alpha diversity) or between samples (beta diversity).</p>
<p>However, diversity is not a physical quantity that can be measured directly, and many different metrics have been proposed to quantify diversity</p>
<p>Species diversity consists of three components: species richness, taxonomic or phylogenetic diversity and species evenness.</p>
<ol style="list-style-type: decimal">
<li><p>Species richness = the number of different species in a community.</p></li>
<li><p>Species evenness = how close the species abundencies in a community are.</p></li>
<li><p>Phylogenetic diversity = how closely related the species in a community are.</p></li>
</ol>
<p>QIIME 2 calculates comprehensive set of metrics for you with one command</p>
<p>Alpha diversity metrics</p>
<p>• Observed Features (a measure of community richness, called “observed OTUs” here for historical reasons)</p>
<p>• Evenness (or Pielou’s Evenness; a measure of community evenness; Pielou, 1966)</p>
<p>• Shannon Diversity Index (a measure of community diversity; Shannon &amp; Weaver, 1949)</p>
<p>• Faith’s Phylogenetic Diversity (a measure of diversity that incorporates phylogenetic relationships between the features; Faith, 1992); this metric is sometimes referred to as PD_whole_tree, but we discourage the use of that name in favor of Faith’s Phylogenetic Diversity or Faith’s PD</p>
<p>Beta diversity metrics</p>
<p>• Jaccard distance (qualitative comparison: # shared OTUs / # union OTUs *100 ; (Jaccard, 1908)</p>
<p>• Bray‐Curtis distance (a quantitative measure of community dissimilarity- includes abundance; Sørensen, 1948)</p>
<p>• unweighted UniFrac distance (a qualitative measure of community dissimilarity- analogous to Jaccard- that incorporates phylogenetic relationships between the features; Lozupone &amp; Knight, 2005); implementation based on Striped UniFrac (McDonald et al., 2018) method</p>
<p>• weighted UniFrac distance (a quantitative measure of community dissimilarity- analogous to Bray-Curtis- that incorporates phylogenetic relationships between the features; Lozupone, Hamady, Kelley, &amp; Knight, 2007); implementation based on Striped UniFrac (McDonald et al., 2018) method.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate diversity metrics</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity core-metrics-phylogenetic <span class="dt">\</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>--i-phylogeny rooted-tree.qza <span class="dt">\</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>--i-table table-or-99.qza <span class="dt">\</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>--p-sampling-depth 1000 <span class="dt">\</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>--output-dir core-metrics-results</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co"># SEVERAL FILES WILL BE GENERATED IN A FOLDER/DIRECTORY TITLED &quot;core-metrics-results&quot;</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 17 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity alpha-group-significance <span class="dt">\</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>--i-alpha-diversity core-metrics-results/faith_pd_vector.qza <span class="dt">\</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/faith-pd-group-significance.qzv</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity alpha-group-significance <span class="dt">\</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>--i-alpha-diversity core-metrics-results/evenness_vector.qza <span class="dt">\</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/evenness-group-significance.qzv</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity alpha-group-significance  <span class="dt">\</span></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>--i-alpha-diversity core-metrics-results/shannon_vector.qza <span class="dt">\</span></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/shannon_group-significance.qzv</span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity beta-group-significance <span class="dt">\</span></span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a>--i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza <span class="dt">\</span></span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a>--m-metadata-column BodySite <span class="dt">\</span></span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/unweighted-unifrac-body-site-significance.qzv <span class="dt">\</span></span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a>--p-pairwise</span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-42"><a href="#cb17-42" aria-hidden="true" tabindex="-1"></a><span class="co">#THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb17-43"><a href="#cb17-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-44"><a href="#cb17-44" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> diversity beta-group-significance <span class="dt">\</span></span>
<span id="cb17-45"><a href="#cb17-45" aria-hidden="true" tabindex="-1"></a>--i-distance-matrix core-metrics-results/unweighted_unifrac_distance_matrix.qza <span class="dt">\</span></span>
<span id="cb17-46"><a href="#cb17-46" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb17-47"><a href="#cb17-47" aria-hidden="true" tabindex="-1"></a>--m-metadata-column Sex <span class="dt">\</span></span>
<span id="cb17-48"><a href="#cb17-48" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/unweighted-unifrac-sex-significance.qzv <span class="dt">\</span></span>
<span id="cb17-49"><a href="#cb17-49" aria-hidden="true" tabindex="-1"></a>--p-pairwise</span>
<span id="cb17-50"><a href="#cb17-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-51"><a href="#cb17-51" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 6 SECONDS TO RUN TO COMPLETION</span></span></code></pre></div>
<p><img src="QIIMEpics/beta_diversity.png" style="width:100.0%" alt="workflow" />
Significance for a beta diversity metric indicates lesser ingroup distances for one or more groups compared to the inter-group distances.</p>
</div>
<div id="pcoa-plot" class="section level2 hasAnchor" number="0.14">
<h2><span class="header-section-number">0.14</span> PCoA plot<a href="#pcoa-plot" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Explore beta diversity metrics. Run any one metric for beta diversity, you do not have to run all the methods. Pick the one that most defines your dataset.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> emperor plot <span class="dt">\</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>--i-pcoa core-metrics-results/unweighted_unifrac_pcoa_results.qza <span class="dt">\</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>--p-custom-axes TreatmentNumber  <span class="dt">\</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/unweighted-unifrac-emperor-TreatmentNumber.qzv</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> emperor plot <span class="dt">\</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>--i-pcoa core-metrics-results/bray_curtis_pcoa_results.qza <span class="dt">\</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>--p-custom-axes TreatmentNumber <span class="dt">\</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>--o-visualization core-metrics-results/bray-curtis-emperor-TreatmentNumber.qzv</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the files to examine in the Qiime viewer</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="at">-P</span> 44111 <span class="at">-r</span> <span class="op">&lt;</span>username<span class="op">&gt;</span>@gateway.training.ncgr.org:~/QIIME2_Analysis/core-metrics-results Desktop/</span></code></pre></div>
</div>
<div id="assigning-taxonomy" class="section level2 hasAnchor" number="0.15">
<h2><span class="header-section-number">0.15</span> Assigning taxonomy<a href="#assigning-taxonomy" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For many experiments, investigators aim to identify the organisms that are present in a sample. E.g., what genera or species are present in my samples? Are there any human pathogens in this patient’s sample? We can do this by comparing our query sequences (i.e., our features, be they ASVs or OTUs) to a reference database of sequences with known taxonomic composition. Simply finding the closest alignment is not really good enough — because other sequences that are equally close matches or nearly as close may have different taxonomic annotations. So we use taxonomy classifiers to determine the closest taxonomic affiliation with some degree of confidence or consensus (which may not be a species name if one cannot be predicted with certainty!), based on alignment, k-mer frequencies, etc.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> <span class="st">&quot;https://data.qiime2.org/2020.6/common/gg-13-8-99-nb-classifier.qza&quot;</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-classifier classify-sklearn <span class="dt">\</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>--i-classifier gg-13-8-99-nb-classifier.qza <span class="dt">\</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>--i-reads rep-seqs-or-99.qza <span class="dt">\</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>--o-classification taxonomy.qza</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#THE ABOVE STEP SHOULD APPROXIMATELY TAKE 6 MINUTES TO RUN TO COMPLETION</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> metadata tabulate <span class="dt">\</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>--m-input-file taxonomy.qza <span class="dt">\</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>--o-visualization taxonomy.qzv</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a><span class="co">#THE ABOVE STEP SHOULD APPROXIMATELY TAKE 5 SECONDS TO RUN TO COMPLETION</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> taxa barplot <span class="dt">\</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>--i-table table-or-99.qza <span class="dt">\</span></span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>--i-taxonomy taxonomy.qza <span class="dt">\</span></span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>--o-visualization taxa-bar-plots.qzv</span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a><span class="co"># THE ABOVE STEP SHOULD APPROXIMATELY TAKE 7 SECONDS TO RUN TO COMPLETION</span></span></code></pre></div>
</div>
<div id="visualizing" class="section level2 hasAnchor" number="0.16">
<h2><span class="header-section-number">0.16</span> Visualizing<a href="#visualizing" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Copy the visualization files</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">scp</span> <span class="at">-P</span> 44111 <span class="op">&lt;</span>USERNAME<span class="op">&gt;</span>@gateway.training.ncgr.org:/home/<span class="op">&lt;</span>USERNAME<span class="op">&gt;</span>/QIIME2_1000_Analysis<span class="dt">\</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>/core-metrics-results/<span class="pp">*</span>.qzv ./</span></code></pre></div>
<p>Open your browser and go to the following site:
<a href="https://view.qiime2.org/" class="uri">https://view.qiime2.org/</a></p>
<p>STEP 12: Differential Abundance - ANCOM - Analysis of Composition of Microbes (Optional)</p>
<p>• Identifies taxa that are present in different abundances across sample groups</p>
<p>• Compares log ratio of the abundance of each taxon to the abundance of all remaining taxa one at a time</p>
<p>• Assumes small percentage (&lt;25%) of features are changing between groups</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Limit ANCOM analysis to samples from a single body site (as many features </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># are changing in abundance across body sites)</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-table filter-samples <span class="dt">\</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>--p-where <span class="st">&quot;BodySite=&#39;Nasal&#39;&quot;</span> <span class="dt">\</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>--o-filtered-table nasal-table.qza</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="ex">=================================================</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="co"># You should run the following command/plugin on your own dataset where you collapse </span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a><span class="co"># features to a chosen taxonomic level before ANCOM </span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co"># (levels are 1-7, but choose something like 5 or 6)</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> taxa collapse <span class="dt">\</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>--i-table nasal-table.qza <span class="dt">\</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>--i-taxonomy taxonomy.qza <span class="dt">\</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>--p-level 6 <span class="dt">\</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>--o-collapsed-table nasal-table-level6.qza</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="ex">==================================================</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="co"># In this step, ANCOM takes logs and adds one count </span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a><span class="co"># to every value (pseudocount)</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> composition add-pseudocount <span class="dt">\</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>--i-table nasal-table.qza <span class="dt">\</span></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>--o-composition-table comp-nasal-table.qza</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a><span class="co">#run ancom </span></span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> composition ancom <span class="dt">\</span></span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a>--i-table comp-nasal-table.qza <span class="dt">\</span></span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a>--m-metadata-file mappingFile.txt <span class="dt">\</span></span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a>--m-metadata-column Treatment <span class="dt">\</span></span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a>--o-visualization ancom-Treatment-nasal.qzv</span></code></pre></div>
<p>EXERCISE: Preparing your raw sequence data and importing -[<a href="https://docs.qiime2.org/2020.11/tutorials/importing/#" class="uri">https://docs.qiime2.org/2020.11/tutorials/importing/#</a>]</p>
<p>When using Earth Microbiome Project (EMP) protocol, you should have access to forward reads, reverse reads (if applicable),</p>
<p>and a barcode list, i.e.:</p>
<ol style="list-style-type: decimal">
<li>forward.fastq.gz</li>
<li>reverse.fastq.gz</li>
<li>barcodes.fastq.gz</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate qiime2</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> emp_example</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> emp_example</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> tools import <span class="dt">\</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>--type EMPPairedEndSequences <span class="dt">\</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>--input-path ./ <span class="dt">\</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>--output-path emp-paired-end-sequences.qza</span></code></pre></div>
</div>
<div id="demultiplexing" class="section level2 hasAnchor" number="0.17">
<h2><span class="header-section-number">0.17</span> Demultiplexing<a href="#demultiplexing" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>To demultiplex sequences we need to know which barcode sequence is associated with each sample. This information is contained in the sample metadata file.</p>
<p>Note: Our TB data was is already demultiplexed.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wget</span> https://data.qiime2.org/2019.1/tutorials/atacama-soils/sample_metadata.tsv</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> demux emp-paired <span class="dt">\</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>--i-seqs emp-paired-end-sequences.qza <span class="dt">\</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>--m-barcodes-file sample_metadata.tsv <span class="dt">\</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>--m-barcodes-column BarcodeSequence <span class="dt">\</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>--o-per-sample-sequences demux.qza <span class="dt">\</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>--p-rev-comp-mapping-barcodes <span class="dt">\</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>--o-error-correction-details demux-details.qza</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="co"># convert the .qza artifact to a .qzv file for visualization</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> demux summarize <span class="dt">\</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>--i-data demux.qza <span class="dt">\</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>--o-visualization demux.qzv</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a><span class="co"># scp the output file to your local machine and view in a web browser.</span></span></code></pre></div>
</div>
<div id="denoising" class="section level2 hasAnchor" number="0.18">
<h2><span class="header-section-number">0.18</span> Denoising<a href="#denoising" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://docs.qiime2.org/2021.2/tutorials/atacama-soils/" class="uri">https://docs.qiime2.org/2021.2/tutorials/atacama-soils/</a></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> dada2 denoise-paired <span class="dt">\ </span>  </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="ex">--i-demultiplexed-seqs</span> demux.qza <span class="dt">\ </span>  </span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="ex">--p-trim-left-f</span> 13 <span class="dt">\</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>--p-trim-left-r 13 <span class="dt">\</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>--p-trunc-len-f 150 <span class="dt">\</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>--p-trunc-len-r 150 <span class="dt">\</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>--o-table table.qza <span class="dt">\</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>--o-representative-sequences rep-seqs.qza <span class="dt">\</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>--o-denoising-stats denoising-stats.qza</span></code></pre></div>
</div>
<div id="chimeric-feature-filtering" class="section level2 hasAnchor" number="0.19">
<h2><span class="header-section-number">0.19</span> Chimeric feature filtering<a href="#chimeric-feature-filtering" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://docs.qiime2.org/2021.2/tutorials/chimera/" class="uri">https://docs.qiime2.org/2021.2/tutorials/chimera/</a></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> vsearch uchime-denovo <span class="dt">\</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>--i-sequences rep-seqs.qza <span class="dt">\</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>--output-dir uchime-dn-out</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="co"># tabulate for visualization</span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> metadata tabulate <span class="dt">\</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>--m-input-file uchime-dn-out/stats.qza <span class="dt">\</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>--o-visualization uchime-dn-out/stats.qzv</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="co"># creates filtered files</span></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-table filter-features <span class="dt">\</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a>--i-table table.qza <span class="dt">\</span></span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>--m-metadata-file uchime-dn-out/nonchimeras.qza <span class="dt">\</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>--o-filtered-table uchime-dn-out/table-nonchimeric-wo-borderline.qza </span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-table filter-seqs <span class="dt">\</span></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>--i-data rep-seqs.qza <span class="dt">\</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>--m-metadata-file uchime-dn-out/nonchimeras.qza <span class="dt">\</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>--o-filtered-data uchime-dn-out/rep-seqs-nonchimeric-wo-borderline.qza </span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="ex">qiime</span> feature-table summarize <span class="dt">\</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>--i-table uchime-dn-out/table-nonchimeric-wo-borderline.qza <span class="dt">\</span></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>--o-visualization uchime-dn-out/table-nonchimeric-wo-borderline.qzv</span></code></pre></div>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/%s",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": false,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
